lib/semantics/check-do.cc

```diff
@@ -64,6 +78,77 @@ class DoConcurrentBodyEnforce {
   template<typename T> bool Pre(const T &) { return true; }
   template<typename T> void Post(const T &) {}
 
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+
+  // C1140 -- Can't deallocate a polymorphic entity in a DO CONCURRENT
+  // Note also that the deallocation of a derived type entity might cause the
+  // invocation of an IMPURE final subroutine.
+  void Post(const parser::BlockConstruct &blockConstruct) {
+    const auto &endBlockStmt{
+        std::get<parser::Statement<parser::EndBlockStmt>>(blockConstruct.t)};
+    const Scope &blockScope{context_.FindScope(endBlockStmt.source)};
+    const Scope &doScope{context_.FindScope(doConcurrentSourcePosition_)};
+    if (DoesScopeContain(&doScope, blockScope)) {
+      for (auto &pair : blockScope) {
+        Symbol &symbol{*pair.second};
+        if (IsPolymorphic(symbol) && IsAllocatable(symbol) &&
+            !symbol.attrs().test(Attr::SAVE)) {
+          SayWithDeclaration(context_, endBlockStmt.source,
+              "Deallocation of a polymorphic entity caused by block"
+              " exit not allowed in DO CONCURRENT"_err_en_US,
+              parser::CharBlock{}, symbol);
+        }
+        // TODO: Check for deallocation of a variable with an IMPURE FINAL
+        // subroutine
+      }
+    }
+  }
+
+  void Post(const parser::AssignmentStmt &stmt) {
+    const auto &variable{std::get<parser::Variable>(stmt.t)};
+    const auto &variableExpr{variable.typedExpr};
+    if (const SomeExpr * expr{GetExpr(variableExpr)}) {
+      const parser::Name &name{GetLastName(variable)};
+      if (const Symbol * symbol{name.symbol}) {
+        if (const std::optional<evaluate::DynamicType> &type{expr->GetType()}) {
+          if (type->IsPolymorphic()) {
+            SayWithDeclaration(context_, variable.GetSource(),
+                "Deallocation of a polymorphic entity caused by "
+                "assignment not allowed in DO CONCURRENT"_err_en_US,
+                parser::CharBlock{}, *symbol);
+          }
```

Comment:
Can you add a comment regarding the check here? Shouldn't there be checks regarding difference in types/size, existing allocation status (that might make it a runtime check though).

Reply:
I added some comments to clarify where deallocation checks are needed.  Assignment statements may require other semantic checks, but this code is only checking for deallocation of polymorphic entities.